File: iOS/Spaces V1.6/ViewModels/OnboardingViewModel.swift
import Combine
import Foundation
import SwiftUI

@MainActor
final class OnboardingViewModel: ObservableObject {
    enum Page: Int, CaseIterable, Identifiable {
        case welcome
        case spaces
        case pings
        case privacy
        case authentication
        case profile

        var id: Int { rawValue }
    }

    static let defaultEmoji = "🧑‍💻"

    @Published var currentPage: Page = .welcome
    @Published var displayName: String = ""
    @Published var emojiAvatar: String = defaultEmoji
    @Published var statusMessage: String = ""
    @Published var isShowingSplash = true

    var canContinueProfile: Bool {
        !displayName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    var displayEmoji: String {
        let trimmed = emojiAvatar.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? Self.defaultEmoji : trimmed
    }

    var currentPageIndex: Int {
        currentPage.rawValue
    }

    var pageCount: Int {
        Page.allCases.count
    }

    var showsNextButton: Bool {
        currentPage != .profile
    }

    func advanceFromSplash() {
        withAnimation(.easeInOut(duration: 0.3)) {
            isShowingSplash = false
        }
    }

    func nextPage() {
        guard let nextPage = Page(rawValue: currentPage.rawValue + 1) else { return }
        withAnimation(.easeInOut(duration: 0.3)) {
            currentPage = nextPage
        }
    }

    func setPage(_ page: Page) {
        withAnimation(.easeInOut(duration: 0.3)) {
            currentPage = page
        }
    }

    func sanitizeEmojiInput() {
        let compact = emojiAvatar.trimmingCharacters(in: .whitespacesAndNewlines)

        if compact.isEmpty {
            emojiAvatar = ""
            return
        }

        if let firstCluster = compact.first {
            emojiAvatar = String(firstCluster)
        }
    }

    func applyAuthenticatedProfileDefaults(displayName: String?) {
        if self.displayName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            self.displayName = displayName?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        }
    }

    func prepareForRequiredProfileCreation(displayName: String?) {
        applyAuthenticatedProfileDefaults(displayName: displayName)
        isShowingSplash = false
        currentPage = .profile
    }
}

File: Android/app/src/main/java/com/arcinteractive/spaces/ui/screens/onboarding/OnboardingViewModel.kt
package com.arcinteractive.spaces.ui.screens.onboarding

import android.icu.text.BreakIterator
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import java.util.Locale

enum class OnboardingPage {
    Welcome,
    Spaces,
    Pings,
    Privacy,
    Authentication,
    Profile
}

data class OnboardingUiState(
    val isShowingSplash: Boolean = true,
    val currentPage: OnboardingPage = OnboardingPage.Welcome,
    val displayName: String = "",
    val emojiAvatar: String = OnboardingViewModel.DEFAULT_EMOJI,
    val statusMessage: String = ""
) {
    val displayEmoji: String = emojiAvatar.ifBlank { OnboardingViewModel.DEFAULT_EMOJI }
    val canContinueProfile: Boolean = displayName.isNotBlank()
}

class OnboardingViewModel : ViewModel() {
    companion object {
        const val DEFAULT_EMOJI = "\uD83E\uDDD1\u200D\uD83D\uDCBB"
    }

    private val _uiState = MutableStateFlow(OnboardingUiState())
    val uiState: StateFlow<OnboardingUiState> = _uiState.asStateFlow()

    fun advanceFromSplash() {
        _uiState.update { it.copy(isShowingSplash = false) }
    }

    fun setCurrentPage(page: OnboardingPage) {
        _uiState.update { it.copy(currentPage = page) }
    }

    fun updateDisplayName(name: String) {
        _uiState.update { it.copy(displayName = name) }
    }

    fun updateEmojiAvatar(emoji: String) {
        val compact = emoji.trim()
        val sanitized = if (compact.isEmpty()) {
            ""
        } else {
            val iterator = BreakIterator.getCharacterInstance(Locale.getDefault())
            iterator.setText(compact)
            val end = iterator.next()
            if (end == BreakIterator.DONE) compact else compact.substring(0, end)
        }

        _uiState.update { it.copy(emojiAvatar = sanitized) }
    }

    fun updateStatusMessage(status: String) {
        _uiState.update { it.copy(statusMessage = status) }
    }

    fun applyAuthenticatedProfileDefaults(displayName: String?) {
        _uiState.update { current ->
            current.copy(
                displayName = if (current.displayName.isBlank()) displayName.orEmpty() else current.displayName
            )
        }
    }

    fun prepareForRequiredProfileCreation(displayName: String?) {
        _uiState.update { current ->
            current.copy(
                isShowingSplash = false,
                currentPage = OnboardingPage.Profile,
                displayName = if (current.displayName.isBlank()) displayName.orEmpty() else current.displayName
            )
        }
    }
}
